### PR TITLE
[AutoDiff] fix ownership instructions (SR-13973)

### PR DESCRIPTION
**Pull request by marcrasi**
I don't really understand what I am doing, but my reasoning is: https://github.com/apple/swift/pull/34693 made `getOwnershipKind()` start returning `None` when the function doesn't have ownership, and this caused Differentiation to insert fewer copy value operations, which caused https://bugs.swift.org/browse/SR-13973. So modify the condition to insert copy value operations when the function doesn't have ownership.

Only the changes on L503 and L1217 are covered by the new test, but I changed the other similar-looking conditions too, because it seems like they probably need the same change.

### CODE REVIEW

**Comment by marcrasi** [comment]
@swift-ci please test

File: test/AutoDiff/validation-test/address_sanitizer.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// RUN: %target-run %t/sr13973
+
+// REQUIRES: executable_test
+// REQUIRES: asan_runtime
+
+import _Differentiation
+
+struct SR13973 {
+  let x: Float = 0
+
```
Comment (gottesmm) [inline test/AutoDiff/validation-test/address_sanitizer.swift:12]: Can I have a SIL test?

Comment (gottesmm) [reply test/AutoDiff/validation-test/address_sanitizer.swift:12]: For changes like this, both are important!

**Review by dan-zheng — APPROVED** [review]
If this change fixes the crash, it nominally LGTM!

I'd like to understand more deeply https://github.com/apple/swift/pull/34693 and the underlying reasons for the crash and why the runtime ASAN error wasn't caught by SIL verification.

File: test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil [hunk 2]
```diff
@@ -139,3 +139,50 @@ bb0(%0 : $Class):
...
+
+// CHECK-LABEL: sil {{.*}} @AD__method_curried__differentiable_curry_thunk_src_0_wrt_0
+// CHECK: bb0([[SELF:%.*]] : $S):
+// CHECK:   [[METHOD:%.*]] = function_ref @method_thin : $@convention(thin) (Float, S) -> Float
+// CHECK:   [[CURRIED:%.*]] = partial_apply [callee_guaranteed] [[METHOD]]([[SELF]]) : $@convention(thin) (Float, S) -> Float
+// CHECK:   [[METHOD_JVP:%.*]] = differentiability_witness_function [jvp] [parameters 0] [results 0] @method_thin : $@convention(thin) (Float, S) -> Float
+// CHECK:   [[CURRIED_JVP:%.*]] = partial_apply [callee_guaranteed] [[METHOD_JVP]]([[SELF]]) : $@convention(thin) (Float, S) -> (Float, @owned @callee_guaranteed (Float) -> Float)
+// CHECK:   [[METHOD_VJP:%.*]] = differentiability_witness_function [vjp] [parameters 0] [results 0] @method_thin : $@convention(thin) (Float, S) -> Float
+// CHECK:   [[CURRIED_VJP:%.*]] = partial_apply [callee_guaranteed] [[METHOD_VJP]]([[SELF]]) : $@convention(thin) (Float, S) -> (Float, @owned @callee_guaranteed (Float) -> Float)
+// CHECK:   strong_retain [[CURRIED]] : $@callee_guaranteed (Float) -> Float
```
Comment (marcrasi) [inline test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil:185]: This `strong_retain` gets added by this PR.

File: test/AutoDiff/validation-test/address_sanitizer.swift (hunk 1 above)
Comment (marcrasi) [reply test/AutoDiff/validation-test/address_sanitizer.swift:12]: I added a FileCheck to test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil. It's pretty much the same thing that I do in this file, except in SIL.

**Review by rxwei — COMMENTED** [review]
Would a more principled fix be enabling OSSA for AD curry and index subset thunks?

File: test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil [hunk 3]
```diff
@@ -139,3 +139,50 @@ bb0(%0 : $Class):
...
+  %2 = function_ref @method_curried : $@convention(thin) (S) -> @owned @callee_guaranteed (Float) -> Float
+  %3 = apply %2(%1) : $@convention(thin) (S) -> @owned @callee_guaranteed (Float) -> Float
+  %4 = differentiable_function [parameters 0] [results 0] %3 : $@callee_guaranteed (Float) -> Float
+  %5 = tuple ()
+  return %5 : $()
+}
+
+// CHECK-LABEL: sil {{.*}} @AD__method_curried__differentiable_curry_thunk_src_0_wrt_0
+// CHECK: bb0([[SELF:%.*]] : $S):
+// CHECK:   [[METHOD:%.*]] = function_ref @method_thin : $@convention(thin) (Float, S) -> Float
```
Comment (gottesmm) [inline test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil:179]: Can you add an ossa test? Doesn't this transform only run on OSSA? I guess not?

File: test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil [hunk 4]
```diff
@@ -139,3 +139,50 @@ bb0(%0 : $Class):
...
+sil @test_curry_thunk : $@convention(thin) (Float, S) -> () {
+bb0(%0 : $Float, %1 : $S):
+  %2 = function_ref @method_curried : $@convention(thin) (S) -> @owned @callee_guaranteed (Float) -> Float
+  %3 = apply %2(%1) : $@convention(thin) (S) -> @owned @callee_guaranteed (Float) -> Float
+  %4 = differentiable_function [parameters 0] [results 0] %3 : $@callee_guaranteed (Float) -> Float
+  %5 = tuple ()
+  return %5 : $()
+}
+
+// CHECK-LABEL: sil {{.*}} @AD__method_curried__differentiable_curry_thunk_src_0_wrt_0
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil]: Done

File: test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil (hunk 3 above)
Comment (marcrasi) [reply test/AutoDiff/SILOptimizer/differentiation_function_canonicalization.sil:222]: Does that mean add a test where the sil functions say [ossa]? Done.

I guess this transform doesn't only run on OSSA...

**Pull request closed without merging**